avocado/core/test.py

```diff
@@ -133,7 +133,156 @@ def str_filesystem(self):
+            test_data_dir = self._get_test_datadir()
+            if test_data_dir is None:
+                return None
+            datadir = astring.string_to_safe_path(self.name.variant)
+            self._cached_variant_datadir = os.path.join(test_data_dir,
+                                                        datadir)
+        return self._cached_variant_datadir
+
+    def _get_data_file(self, filename, must_exist):
+        datadir = self._get_file_datadir()
+        if datadir is None:
+            return None
+        file_filename = os.path.join(self.datadir, filename)
+        if not must_exist:
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, file_filename,
+                           "to be found at file-level data dir")
+            return file_filename
+        else:
+            if os.path.exists(file_filename):
+                self.log.debug(self.GET_DATA_LOG_FMT, filename, file_filename,
+                               "found at file-level data dir")
+                return file_filename
+
+    def _get_data_test(self, filename, must_exist):
+        if self.datadir is None or not os.path.isdir(self.datadir):
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, None,
+                           "no data dir exists")
+            return None
+
+        test_data_dir = self._get_test_datadir()
+        test_filename = os.path.join(test_data_dir, filename)
+        if not must_exist:
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, test_filename,
+                           "to be found at test-level data dir")
+            return test_filename
+        else:
+            if os.path.exists(test_filename):
+                self.log.debug(self.GET_DATA_LOG_FMT, filename, test_filename,
+                               "found at test-level data dir")
+                return test_filename
+
+    def _get_data_variant(self, filename, must_exist):
+        if self.name.variant is None:
+            return None
+        variant_data_dir = self._get_variant_datadir()
+        if variant_data_dir is None:
+            return None
+        variant_filename = os.path.join(variant_data_dir, filename)
+        if not must_exist:
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, variant_filename,
+                           "to be found at variant-level data dir")
+            return variant_filename
+        else:
+            if os.path.exists(variant_filename):
+                self.log.debug(self.GET_DATA_LOG_FMT, filename,
+                               variant_filename, "found at variant-level data dir")
+                return variant_filename
+
+    def get_data(self, filename, source=None, must_exist=True):
+        if source is None:
+            sources = reversed(self.DATA_SOURCES)
+        else:
+            self._check_valid_data_source(source)
+            sources = [source]
+        for attempt_source in sources:
+            get_callable = "_get_data_%s" % attempt_source
```

Comment:
This thing encourage people to just create another `self._get_data_...` methods, which is the thing I dislike the most about this implementation (which means NACK from my side). Yes, you say it's one of the implementation and it's not public API, but I disagree about that. This is the implementation which is part of ever INSTRUMENTED test, so this is what people get and should use. If you want them to define other sources, this is what they will be using and any change would break compatibility.

Would you have anything against a simple `self._data_dirs` list, which can contain:

1. actual paths
2. methods which on call replace themselves with actual path (lazy init)

As for the `sources`, as I mentioned I'd like to avoid that in the first version, but if you still see this coming you can prepare for it by storing not just paths, but also tuples (or struct with `.name` and `.path`) which would be used to filter the results.

As for the extensibility, one would simply modify:

```python
def __init__(self, ...):
    super(...)
    self._data_dirs = []
```

Here the names would actually come handy as you might want to remove just some of them based on the name:

```python
def __init__(self, ...):
    super(...)
    self._data_dirs = [_ for _ in self._data_dirs if _.name != "test"]
```

Reply:
I have an over-night update, if you want to keep the implementation, I'd be fine with `Test.data_dirs` object, which would basically consist of the `TestData` class. The necessary changes would be:

1. The `TestData` class would have to be public API (at least the `BaseTestData`), either in `avocado` or in `data_structures`. On `Test.__init__` this class would be instantiated, unless already set by the ancestor class.
2. Modification to `data_dirs` would be done via manual instantiating of `BaseTestData` (or `TestData`) similarly to how it works here, basically on `TestData.__init__` one would tweak the data sources, or simply inherit from it and modify just the necessary parts.
3. the `Test.get_data()` would basically point to `Test.data_dirs.get_data()`

I'd still not love such implementation, but I'd not NACK it as the main reason is polluting the `avocado` class.